src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,163 @@
+            this.prev = prev;
+        }
+    }
+
     @Override
-    public void add(T value) {
+    public void add(T element) {
+        linkLast(element);
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T element, int index) {
+        isInvalidIndex(index);
+        if (index == size) {
+            linkLast(element);
+        } else {
+            linkBefore(element, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el : list) {
+            linkLast(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexPos(index);
+        return findNodeByIndex(index).element;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T set(T element, int index) {
+        checkIndexPos(index);
+        Node<T> x = findNodeByIndex(index);
+        T original = x.element;
+        x.element = element;
+        return original;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexPos(index);
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> node = first; node != null; node = node.next) {
+                if (node.element == null) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> node = first; node != null; node = node.next) {
+                if (object.equals(node.element)) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current = first;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private void isInvalidIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Index is invalid");
+        }
+    }
+
+    private void checkIndexPos(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException("Index is invalid");
+        }
+    }
+
+    private void linkBefore(T e, int index) {
+        Node<T> midNode = findNodeByIndex(index);
+        Node<T> someNode = midNode.prev;
+        Node<T> newNode = new Node<>(someNode, e, midNode);
+        midNode.prev = newNode;
+        if (someNode == null) {
+            first = newNode;
+        } else {
+            someNode.next = newNode;
+        }
+        size++;
+    }
+
+    void linkLast(T e) {
+        Node<T> l = last;
```

Comment:
Variables should be named more informatively

Reply:
fixed